File: backend/internal/usecase/member_usecase.go
package usecase

import (
	"context"

	"healthfamily/internal/domain"
	"healthfamily/internal/domain/entity"
	"healthfamily/internal/domain/repository"
)

// MemberUsecase はメンバー管理のビジネスロジック
type MemberUsecase struct {
	members repository.MemberRepository
}

func NewMemberUsecase(members repository.MemberRepository) *MemberUsecase {
	return &MemberUsecase{members: members}
}

// ensureOwner は指定メンバーが userID の所有であることを確認する
func (uc *MemberUsecase) ensureOwner(ctx context.Context, userID, memberID string) (*entity.Member, error) {
	m, err := uc.members.FindByID(ctx, memberID)
	if err != nil {
		return nil, err
	}
	if m == nil {
		return nil, domain.NewNotFound("メンバー")
	}
	if m.UserID != userID {
		return nil, domain.NewForbidden("このメンバーにアクセスする権限がありません")
	}
	return m, nil
}

func (uc *MemberUsecase) List(ctx context.Context, userID string) ([]entity.Member, error) {
	return uc.members.List(ctx, userID)
}

// ListSummary はメンバー＋薬数の集計を返す（N+1回避）
func (uc *MemberUsecase) ListSummary(ctx context.Context, userID string) ([]entity.MemberSummary, error) {
	return uc.members.ListSummary(ctx, userID)
}

func (uc *MemberUsecase) Get(ctx context.Context, userID, memberID string) (*entity.Member, error) {
	return uc.ensureOwner(ctx, userID, memberID)
}

func (uc *MemberUsecase) Create(ctx context.Context, in repository.CreateMemberInput) (*entity.Member, error) {
	if in.MemberType == "" {
		in.MemberType = "human"
	}
	return uc.members.Create(ctx, in)
}

func (uc *MemberUsecase) Update(ctx context.Context, userID, memberID string, in repository.UpdateMemberInput) (*entity.Member, error) {
	if _, err := uc.ensureOwner(ctx, userID, memberID); err != nil {
		return nil, err
	}
	return uc.members.Update(ctx, memberID, in)
}

func (uc *MemberUsecase) Delete(ctx context.Context, userID, memberID string) error {
	if _, err := uc.ensureOwner(ctx, userID, memberID); err != nil {
		return err
	}
	return uc.members.Delete(ctx, memberID)
}

File: backend-java/src/main/java/app/healthfamily/usecase/member/MemberUseCase.java
package app.healthfamily.usecase.member;

import app.healthfamily.domain.member.Member;
import app.healthfamily.domain.member.MemberRepository;
import app.healthfamily.domain.member.MemberType;
import app.healthfamily.domain.member.PetType;
import app.healthfamily.domain.shared.AppZone;
import app.healthfamily.domain.shared.DomainException;
import java.time.LocalDate;
import java.util.List;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * メンバーの登録と参照。
 *
 * <p>種別と動物種別の整合や未来日の検証は集約が持つ。ここは読み書きの手順だけを担う。
 */
@Service
public class MemberUseCase {

    private final MemberRepository members;
    private final AppZone zone;

    public MemberUseCase(MemberRepository members, AppZone zone) {
        this.members = members;
        this.zone = zone;
    }

    @Transactional
    public View register(RegisterCommand command) {
        LocalDate today = zone.today();
        Member member =
                Member.builder()
                        .id(UUID.randomUUID().toString())
                        .userId(command.userId())
                        .name(command.name())
                        .type(MemberType.fromCode(command.memberType()))
                        .petType(command.petType() == null ? null : PetType.fromCode(command.petType()))
                        .birthDate(command.birthDate())
                        .photoUrl(command.photoUrl())
                        .notes(command.notes())
                        .build(today);

        members.create(member);
        return View.of(member, today);
    }

    @Transactional(readOnly = true)
    public List<View> list(String userId) {
        LocalDate today = zone.today();
        return members.listByUser(userId).stream().map(m -> View.of(m, today)).toList();
    }

    @Transactional(readOnly = true)
    public View get(String userId, String memberId) {
        Member member =
                members.findById(memberId).orElseThrow(() -> DomainException.notFound("メンバー"));
        member.requireOwnedBy(userId);
        return View.of(member, zone.today());
    }

    /** @param age 生年月日から算出した満年齢。未設定なら null */
    public record View(
            String id,
            String name,
            String memberType,
            String petType,
            LocalDate birthDate,
            Integer age,
            String photoUrl,
            String notes) {

        static View of(Member m, LocalDate today) {
            return new View(
                    m.id(),
                    m.name(),
                    m.type().code(),
                    m.petType().map(PetType::code).orElse(null),
                    m.birthDate().orElse(null),
                    m.ageAt(today).orElse(null),
                    m.photoUrl().orElse(null),
                    m.notes().orElse(null));
        }
    }

    public record RegisterCommand(
            String userId,
            String name,
            String memberType,
            String petType,
            LocalDate birthDate,
            String photoUrl,
            String notes) {}
}
